src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.ResolvedDependencyGraphItem.cs

```diff
@@ -35,17 +35,29 @@ public class ResolvedDependencyGraphItem
             /// <param name="item">The <see cref="GraphItem{TItem}" /> with a <see cref="RemoteResolveResult" /> of the resolved library.</param>
             /// <param name="dependencyGraphItem">The <see cref="DependencyGraphItem" /> that is being resolved.</param>
             /// <param name="indexingTable">The <see cref="DependencyGraphItemIndexer" /> to use when indexing dependencies.</param>
+            /// <param name="parents">An optional <see cref="HashSet{T}" /> containing parent <see cref="LibraryRangeIndex" /> values.</param>
             public ResolvedDependencyGraphItem(
                 GraphItem<RemoteResolveResult> item,
                 DependencyGraphItem dependencyGraphItem,
-                DependencyGraphItemIndexer indexingTable)
+                DependencyGraphItemIndexer indexingTable,
+                HashSet<LibraryRangeIndex>? parents = null)
             {
                 Item = item;
 
                 LibraryDependency = dependencyGraphItem.LibraryDependency;
                 LibraryRangeIndex = dependencyGraphItem.LibraryRangeIndex;
                 Path = dependencyGraphItem.Path;
 
+                if (parents != null)
+                {
+                    Parents = parents;
+                }
+                else if (dependencyGraphItem.IsCentrallyPinnedTransitivePackage && !dependencyGraphItem.IsRootPackageReference)
+                {
+                    // Keep track of parents if the item is transitively pinned
+                    Parents = new HashSet<LibraryRangeIndex> { dependencyGraphItem.Parent };
+                }
```

Comment:
Question: Is it intentional that we want the default case to have Parents being null? Would an empty Hashset fit the design better?

Reply:
Yes, this is an optimization at the moment.  If the project is not using central transitive pinning, the parents aren't needed.  If the project is using central transitive pinning, we need to know the parents to calculate the union of all of the IncludeAssets.

So at the moment, we don't new up the Parents property or track the parents unless central transitive pinning is being used, mostly just to cut down on allocations, list resizing, etc.  Moving the logic here is technically an improvement because its only created in this code path, and then everything else just uses the `?` operator rather than checking if central transitive pinning is enabled.  I'm open to more suggestions, the design right now is the most optimal but as you point out, not very clear.

Reply:
@jeffkl @jebriede please follow up offline if necessary. This PR is needed to fix a high priority perf regression